src/algebra/power_mod.lean

```diff
@@ -0,0 +1,135 @@
+/-
+Copyright (c) 2019 Neil Strickland. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Neil Strickland
+
+Given a monoid `M` and an element `g ∈ M` satisfying `gⁿ = 1`, we
+can interpret `gⁱ` for elements `i ∈ ℤ/n`.  We give several
+different Lean versions of this idea.  Some of them involve
+the type `zmod n`, which is Lean's implementation of `ℤ/n`.
+Others involve the relation `a ≡ b [MOD n]` on `ℕ`, or the
+relation `a ≡ b [ZMOD n]` on `ℤ`.
+-/
+
+import data.fintype algebra.group_power data.zmod.basic
+
+variable n : ℕ+
+section pow_mod
+
+variables {M : Type*} [monoid M] {g : M} (hg : g ^ n.val = 1)
+include hg
+
+lemma pow_exponent (m : ℕ) : (g ^ m) ^ n.val = 1 :=
+by {rw[← pow_mul, mul_comm, pow_mul, hg, one_pow]}
+
+lemma pow_nat_mod (i : ℕ) : g ^ i = g ^ (i % n) :=
+calc
+  g ^ i = g ^ (i % n.val + n.val * (i / n.val)) :
+   congr_arg (has_pow.pow g) (nat.mod_add_div i n.val).symm
```

Comment:
Can't you prove this `by rw [← nat.mod_add_div]`?

Reply:
I think that that is what I tried first, and it does not work because it rewrites all three instances of `i`. 
 However, it does work if you remove the left arrow, and that is certainly better.